python/lsst/daf/butler/core/dimensions/_schema.py

```diff
@@ -170,7 +170,7 @@ def __init__(self, element: DimensionElement):
         # don't want *recursive* implied dependencies.
         self.implied = NamedValueSet()
         for dimension in element.implied:
-            fieldSpec = addDimensionForeignKey(self._tableSpec, dimension, primaryKey=False, nullable=True)
+            fieldSpec = addDimensionForeignKey(self._tableSpec, dimension, primaryKey=False, nullable=False)
```

Comment:
Could we also add explicit `nullable=False` to required elements above (line 155)? They are part of primary key so they are implicitly non-NULL, and it looks like sqlalchemy is OK with the column definition which is both PK and nullable, but this is not consistent with the schema when it is reflected back from database. This is not a schema change, of course, just some inconsistency that I discovered in looking at checksusm.

Reply:
Setting `primaryKey=True` does not sett `nullable=False` automatically? @TallJimbo made the above change. If he agrees I can add `nullable=False` to line 155.

Reply:
I had just assumed sqlalchemy would ignore `nullable` for primary keys at a lower level.  If that's not true at some level of interest to us as @andy-slac says, yeah, we should change it - though maybe the `FieldSpec` constructor or `__post_init__` (it's a dataclass) would be a more thorough way to do that.

Reply:
It's a bit complicated, in sqlalchemy indeed setting `primaryKey=True` does not automatically set `nullable=False`, so when you inspect sqlalchemy `Column` you can see the definitions where both `PK` and `nullable` are true. When a table is created in the database then database of course does the right thing. And when you reflect schema from the database the PK columns always have `nullable=False`. I think it would better if reflected definitions and definitions that we construct ourselves were matching each other (but there are other things that don't match of course). And indeed, `FieldSpec` is probably a better place to enforce that consistently.

Reply:
@TallJimbo - one more issue with this update - I tried to implement migration script for dimensions manager (from 6.0.1 to 6.0.2) by altering the columns that correspond to implied dimensions to be `NOT NULL` and using the copy of the `main_20210215` repository. My script failed because `visit` table has NULLs in `physical_filter` column for some records (all come from LATISS). Not sure what to do about it.

Reply:
Hmm. This is probably some very old LATISS data with a broken translator. It's all going to be unusable if it can't be linked to a physical filter. Maybe we need to recalculate the exposure records for them.

Reply:
LATISS has an "unknown~unknown" `physical_filter` we could update these rows with (same with `exposure`), and I imagine any other instrument for which this is a possibility has a similar sentinel value, and it might be best to just do that now to get us through the migration first.  But we should record the IDs of updated `visits` and `exposures`, too, so we can come back later and check them with the latest translator and visit-definition code.

Reply:
`exposure` does not have any NULLs, they only appear in `visit`. There are 182 of them. I made a page with the CSV file attached which lists instrument and visit ID of those rows: https://confluence.lsstcorp.org/display/~salnikov/NULL+physical_filter+values+in+visit+table

I'll update things in my local copy to continue with my tests and we can update central databases later.

Reply:
That's a bit of a bug then because the vist filter should match the exposure one. Maybe we need to re-run define-visits in update mode for them.

Reply:
So FieldSpec defaults to nullable=True but  addDimensionForeignKey defaults to nullable=False and that forces the FieldSpec to be nullable=False regardless (since FieldSpec has no way of stopping someone changing nullable after it's constructed). I don't know if we want to add special logic to FieldSpec to disallow setting nullable=False if primaryKey==True but I can at least fix addDimensionForeignKey to make it use a dynamic default based on primaryKey.

Reply:
I looked at the physical filters for exposures associated with visits that have NULL physical_philter. Those visits have multiple exposures with different physical filters, e.g.:
```
instrument|visit           |exposure     |visit_system|visit_filter|exposure_filter        
----------+----------------+-------------+------------+------------+-----------------------
LATISS    |1911254347800000|2021012000118|           1|            |empty~ronchi90lpmm     
LATISS    |1911254347800000|2021012000119|           1|            |quadnotch1~ronchi90lpmm
LATISS    |1911254347800000|2021012000120|           1|            |BG40~empty             
LATISS    |1911254347800000|2021012000121|           1|            |RG610~empty            
```

Reply:
I just ran the test suite where I added a RuntimeError if the fieldspec coming out of addDimensionForeignKey had nullable=True and primaryKey=True and all tests passed.

Reply:
Ah. Okay. That's us suffering from the current visit_system=1 definition that uses the exposure.group. It means that they had one observing script but kept changing the filter. When define-visits runs, it drops the entry for filter because it changed during the visit. In the new schema change world the default visit definition will make a filter change impossible.

Reply:
Anyhow, adding `unknown~unknown` for the visit filters there is fine as a hack.